Fix panic in distributor due to early cleanup

replay opened the pull request:
Fixes: https://github.com/grafana/mimir/issues/2266

Edit 2022-07-22:
Most of the code in `pkg/distributor/forwarding` has been re-written so looking at the diff between the old and the new code is probably not very helpful when reviewing, I recommend looking at the new implementation as if it were completely new.

In a conversation via DMs with @bboreham we concluded that it would be better to keep the forwarding logic as isolated from the distributor's `PushWithCleanup()` as possible, this should help to ensure that there are no race conditions or pool usage bugs.

bboreham commented:
This feature seems incredibly complicated to implement, thus hard to see if there are any remaining bugs.

Conceptually it seems to me it might work better as a proxy in front of the distributor, either a separate process or a wrapper function in the code which strips out those metrics which are to be forwarded and passes the remainder to the function which sends to ingesters.

bboreham commented:
Some good thoughts there; I wanted to pick up one point:

> The cleanup in the current PushWithCleanup could then be implemented as a middleware like:
...
> 	defer mimirpb.ReuseSlice(req.Timeseries)

The "cleanup", which is to reuse memory and avoid garbage-collection, has to wait until all calls from distributor to ingester have finished.  This is after the point where a `defer` would run, because we return after getting a majority (e.g. 2 out of 3) of responses, and some calls are still running.

pstibrany commented:
This discussion is somewhat related to discussion we have in https://github.com/grafana/mimir/pull/2419.

dimitarvdimitrov commented:
Pushed some example middleware on a different branch just to demonstrate what I meant 435daaf6b23971f6ce8915ea9266d13492b60a14 It might not compile, also not sure where to put the ctx yet

replay commented:
This PR has too much history which goes back and forth between different approaches so the git history isn't useful for reviewing anyway, I'll squash & rebase.

56quarters commented on pkg/distributor/forwarding/forwarding.go:
```diff
@@ -96,185 +74,260 @@ func NewForwarder(reg prometheus.Registerer, cfg Config) Forwarder {
-		return true
-	}
-	r.samples.Add(float64(len(sample.Samples)))
+		if forwardingTarget != "" {
```
Could this be extracted to a method on the `tsByTargets` type? This code is a bit dense but seems to end up appending timeseries by the forwarding target.

replay replied:
good idea, done here: d5b6113f4

56quarters commented on pkg/distributor/forwarding/forwarding.go:
```diff
@@ -96,185 +74,260 @@ func NewForwarder(reg prometheus.Registerer, cfg Config) Forwarder {
 
-	return rule.Ingest
-}
+		if ingest {
```
I'm assuming that we do this work managing indexes for performance reasons? Are there any numbers around the allocations this saves vs the naive case? It seems rather complex to avoid allocating an extra slice of series to ingest.

replay replied:
You are correct that this is for performance reasons. I simplified the logic a bit and also commented the relevant numbers here:

https://github.com/grafana/mimir/pull/2295/#discussion_r931610962

56quarters commented on pkg/distributor/forwarding/forwarding.go:
```diff
@@ -96,185 +74,260 @@ func NewForwarder(reg prometheus.Registerer, cfg Config) Forwarder {
+		f.newRequest(ctx, endpoint, ts, &requestWg, errCh)
 	}
+
+	go func() {
```
I must be missing something. I don't understand why this isn't just code outside of a goroutine.

replay replied:
without the goroutine the call to `.Forward()` would block until the forwarding requests have completed because of the `requestWg.Wait()`. Running this in a go routine `.Forward()` returns early, allowing the distributor to submit the write requests to the ingesters concurrently with the forwarding requests being executed, once the write requests to the ingesters have completed the distributor waits for the `errCh` to close in order to block until the forwarding requests are also done and handle potential errors. 

I'll add a comment explaining that.

replay replied:
Explained here: 0f1f750a4

56quarters commented on pkg/distributor/forwarding/forwarding.go:
```diff
@@ -96,185 +74,260 @@ func NewForwarder(reg prometheus.Registerer, cfg Config) Forwarder {
 
-			nonRecoverable = httpgrpc.Errorf(http.StatusBadRequest, "endpoint %s: %s", endpoint, err.Error())
-		}
+// newRequest launches a new forwarding request and sends it to a worker via a channel.
```
`newRequest` makes me think this is going to create a new request object and return it. Maybe something like `submitForwardingJob()` ?

replay replied:
good point, fixed here: 83357caff44df54ad2a46680eecfe262ed735a3f

colega commented:
I see that `forwardSamples` is called only from `Push`, but the method that API uses is `PushWithCleanup`.

So this basically does not forward samples ingested through the API (`Push` is only used by the ruler as far as I can see).

Is that a bug?

colega commented on pkg/distributor/forwarding/forwarding.go:
```diff
@@ -96,185 +74,260 @@ func NewForwarder(reg prometheus.Registerer, cfg Config) Forwarder {
+	req.exemplars = f.exemplarsTotal
+	req.latency = f.requestLatencyHistogram
+
+	f.reqCh <- req
```
This should respect the `ctx` and stop blocking if context was cancelled.

replay replied:
good catch, thx. 
Fixed in f6265e471

colega commented on pkg/distributor/distributor.go:
```diff
@@ -908,18 +893,41 @@ func (d *Distributor) PushWithCleanup(ctx context.Context, req *mimirpb.WriteReq
+
+	resp, err := d.PushWithCleanup(ctx, req, cleanup)
+
+	errs := []error{err}
+	for err := range forwardingErrCh {
+		errs = append(errs, err)
+	}
```
I don't know where to write this comment, I think this is the best place: if all workers are busy or in a deadlock or whatever, this will block forever even if the context was canceled.

replay replied:
good catch! this is fixed now:

https://github.com/grafana/mimir/blob/1373bff99be70bbb9433ce26b9b53aac6eb28b21/pkg/distributor/distributor.go#L630-L644

56quarters commented on pkg/distributor/forwarding/forwarding.go:
```diff
@@ -96,185 +74,260 @@ func NewForwarder(reg prometheus.Registerer, cfg Config) Forwarder {
-		errors:   r.errorsTotal,
-		samples:  r.samplesTotal,
-		latency:  r.requestLatencyHistogram,
+	var requestWg sync.WaitGroup
```
If I'm reading this code right:
* `forwarder` creates a pool of worker goroutines
* We're submitting jobs for them to run every time `Forward` is called and giving the caller a channel to read errors.

Seems like a job pool might be something we already have in some utility code or dskit. Offhand, there's [this method](https://github.com/grafana/dskit/blob/main/concurrency/runner.go#L87-L115) but it doesn't enforce a max number of goroutines across multiple calls to `Forward` at the same time.

replay replied:
> If I'm reading this code right:
>
> * forwarder creates a pool of worker goroutines
> * We're submitting jobs for them to run every time Forward is called and giving the caller a channel to read errors.

Correct

> Seems like a job pool might be something we already have in some utility code or dskit. Offhand, there's [this method](https://github.com/grafana/dskit/blob/main/concurrency/runner.go#L87-L115) but it doesn't enforce a max number of goroutines across multiple calls to Forward at the same time.

You're right that this might likely be a pattern which is repeated somewhere else, so it would probably make sense to have support for this in dskit. Since we're under time pressure to get this PR merged I think I shouldn't add the support to dskit now and then update dskit in this PR because then there's a risk about further delays in the dskit PR, but I'd be happy to later implement a generic way for this worker pool pattern in dskit and migrate the forwarder to use it.

We could solve this by using dskit's existing `ForEachJob()`, but as you already mentioned this would mean that there won't be any concurrency limit on the forwarding requests which could lead to unbounded memory growth if a forwarding target is slow, plus this code is also quite sensitive to allocations because we're expecting to ship rather large amounts of data to forwarding targets and having a static pool of workers should lead to less allocations.

56quarters replied:
:+1: fine to do this later

colega commented on pkg/distributor/forwarding/forwarding.go:
```diff
@@ -96,185 +74,260 @@ func NewForwarder(reg prometheus.Registerer, cfg Config) Forwarder {
+}
 
-		tsByEndpoint: make(map[string]*[]mimirpb.PreallocTimeseries),
+func (f *forwarder) Forward(ctx context.Context, rules validation.ForwardingRules, in []mimirpb.PreallocTimeseries) (TimeseriesCounts, []mimirpb.PreallocTimeseries, chan error) {
```
I would like to see the behavior of this function documented, in order to understand the code and the usage better.

Currently it looks like it will enqueue requests that eventually will return an error, if any, through the returned chan. 

This allows the caller to do other operations, i.e. `PushWithCleanup`, while this is forwarding.

However, I also can see that this _will_ block if too many requests are already enqueued, so its behavior isn't inconsistent and I don't know what we should expect.

replay replied:
Makes sense, how is this:

https://github.com/grafana/mimir/blob/56f1d28a6f28782e01f135630bbe3dff64c20b83/pkg/distributor/forwarding/forwarding.go#L127-L140

colega replied:
Thank you!

colega commented on pkg/distributor/forwarding/pools.go:
```diff
@@ -0,0 +1,127 @@
+	p.getBytesReader = getBytesReader(&p.bytesReader)
+	p.putBytesReader = putBytesReader(&p.bytesReader)
+	p.getTsByTargets = getTsByTargets(&p.tsByTargets)
+	p.putTsByTargets = putTsByTargets(&p.tsByTargets)
```
Not a blocking nitpick, but we're already compiling with go 1.18, so if we enabled it in `go.mod` this is a good candidate for generics, as we can cover everything with just two functions:

```go
func getter[T any](pool *sync.Pool) func() T {
	return func() T {
		return pool.Get().(T)
	}
}

func putter[T any](pool *sync.Pool) func(T) {
	return func(val T) {
		pool.Put(val)
	}
}
```

And change this to:
```suggestion
	p.getProtobuf = getter[*[]byte](&p.protobuf)
	p.putProtobuf = putter[*[]byte](&p.protobuf)
	p.getSnappy = getter[*[]byte](&p.snappy)
	p.putSnappy = putter[*[]byte](&p.snappy)
	p.getReq = getter[*request](&p.request)
	p.putReq = putter[*request](&p.request)
	p.getBytesReader = getter[*bytes.Reader](&p.bytesReader)
	p.putBytesReader = putter[*bytes.Reader](&p.bytesReader)
	p.getTsByTargets = getter[tsByTargets](&p.tsByTargets)
	p.putTsByTargets = putter[tsByTargets](&p.tsByTargets)
```

replay replied:
I love that idea, but I don't want to change the version in `go.mod` within this request. I'll make a separate one and then once it's merged I'll rebase this one.

replay replied:
This is so much better ❤️ 

https://github.com/grafana/mimir/blob/81aefe52b7c72976c8d7b7c4f0c56362ad7411d7/pkg/distributor/forwarding/pools.go#L48-L61

colega commented on pkg/distributor/forwarding/forwarding.go:
```diff
@@ -96,185 +74,260 @@ func NewForwarder(reg prometheus.Registerer, cfg Config) Forwarder {
-		// to anywhere but we should still send it to the Ingesters.
-		return true
-	}
+// forwardingEndpointAssigner returns a func which takes a timeseries and assigns it endpoints to which it should be forwarded.
```
Outdated comment.

replay replied:
thx! 
fixed in 1f86587e23e0b3fea2bc0137c3514dc7d42021e9

colega commented on pkg/distributor/forwarding/forwarding.go:
```diff
@@ -96,185 +74,260 @@ func NewForwarder(reg prometheus.Registerer, cfg Config) Forwarder {
+
+		metric, err := extract.UnsafeMetricNameFromLabelAdapters(ts.Labels)
+		if err != nil {
+			// Can't check whether a timeseries should be forwarded if it has no metric name.
+			// Send it to the Ingesters and don't forward it.
+			ingest = true
```
Whould we log this?

colega replied:
Oh okay, I see that the error can _only_ be "no metric name label". IMO it would be clearer if that function returned a bool instead.

colega commented on pkg/distributor/forwarding/forwarding.go:
```diff
@@ -96,185 +74,260 @@ func NewForwarder(reg prometheus.Registerer, cfg Config) Forwarder {
 
-	return rule.Ingest
-}
+		if ingest {
+			if tsSliceWriteIdx != tsSliceReadIdx {
+				// Swap the timeseries at the reading and writing indices,
+				// later we'll return all timeseries beyond the write index to the pool.
+				tsSlice[tsSliceWriteIdx], tsSlice[tsSliceReadIdx] = tsSlice[tsSliceReadIdx], tsSlice[tsSliceWriteIdx]
+			}
+			tsSliceWriteIdx++
+		} else {
+			notIngestedCounts.count(ts)
+		}
+	}
+
+	// Truncate the toIngest slice to the index up to which we wrote TimeSeries data into it,
+	// all the TimeSeries objects beyond the write index must be returned to the pool.
+	for _, ts := range tsSlice[tsSliceWriteIdx:] {
+		f.pools.putTs(ts.TimeSeries)
+	}
+	tsSlice = tsSlice[:tsSliceWriteIdx]
```
Even though all this code seems ok, IMO its complexity doesn't pay it's performance. I had to stop and think deeply about this `for` loop modifying the loop inside.

I'd consider just building a new slice with series to ingest, removing them from the original one, and then cleaning up the original one.
Getting a slice from the pool is almost zero price since we have lots of them already, so it won't be much worse in performance.

replay replied:
I agree that unfortunately this is relatively complicated. The alternative which you're suggesting is a bit more complex than just getting a new slice from the pool and returning the old one because we also need to ensure that all of the `*mimirpb.TimeSeries` objects within the slice get returned to the pool and only once. This is simple if we deep copy them into the new slice, but I would prefer to avoid deep copying whenever possible because that's relatively expensive.

I implemented this here: 05c7e94e808220b52a0fb58db8c61ec5020db1e0

This is how the benchmarks look like before/after (with `-test.benchtime=3s`):

```
$ benchstat -delta-test none ./old ./new
name                                                      old time/op    new time/op    delta
RemoteWriteForwarding/forwarding_and_ingesting-4            1.63ms ± 0%    1.67ms ± 0%   +2.52%
RemoteWriteForwarding/forwarding_and_no_ingesting-4         1.65ms ± 0%    1.74ms ± 0%   +5.55%
RemoteWriteForwarding/no_forwarding_and_only_ingesting-4    1.38ms ± 0%    1.42ms ± 0%   +3.25%

name                                                      old alloc/op   new alloc/op   delta
RemoteWriteForwarding/forwarding_and_ingesting-4            2.74kB ± 0%    2.77kB ± 0%   +1.09%
RemoteWriteForwarding/forwarding_and_no_ingesting-4         3.02kB ± 0%    3.85kB ± 0%  +27.46%
RemoteWriteForwarding/no_forwarding_and_only_ingesting-4      183B ± 0%      209B ± 0%  +14.21%

name                                                      old allocs/op  new allocs/op  delta
RemoteWriteForwarding/forwarding_and_ingesting-4              29.0 ± 0%      30.0 ± 0%   +3.45%
RemoteWriteForwarding/forwarding_and_no_ingesting-4           30.0 ± 0%      33.0 ± 0%  +10.00%
RemoteWriteForwarding/no_forwarding_and_only_ingesting-4      4.00 ± 0%      5.00 ± 0%  +25.00%
```

Obviously it's hard to define how many allocations are "too much" relative to complexity of the code :) But I'd be fine to leave it like this.

colega commented on pkg/distributor/forwarding/forwarding.go:
```diff
@@ -96,185 +74,260 @@ func NewForwarder(reg prometheus.Registerer, cfg Config) Forwarder {
-	if len(r.tsByEndpoint) == 0 {
-		close(errCh)
-		return errCh
+	if cap(ts) > len(ts) {
+		ts = ts[:newPos+1]
+	} else {
+		ts = append(ts, mimirpb.PreallocTimeseries{})
 	}
```
Can you explain why are we doing this?

replay replied:
sorry, now that I look at it again I think it makes no sense, removing it. Thanks for catching.

colega commented on pkg/distributor/forwarding/forwarding.go:
```diff
@@ -290,20 +343,30 @@ func (r *request) sendToEndpoint(ctx context.Context, endpoint string, ts []mimi
+	r.pools.putTsSlice(r.ts.ts)
+
+	// Ensure we don't modify a request property after returning it to the pool by calling .Done() on the wait group.
+	wg := r.requestWg
+	r.requestWg = nil
+
+	r.pools.putReq(r)
 
-	r.tsByEndpoint = nil
+	wg.Done()
```
Why can't we just do this?
```go
r.requestWg.Done()
r.pools.putReq(r)
```

replay replied:
Simply for testability :)

If the object would be returned to the pool after the request wait group is done then there is no way to ensure that we only validate the pool usage after all objects have been returned to the pool. If you have a better idea how to handle that then please let me know.

For now I'll add a comment mentioning this.

replay replied:
added this comment: 59784456a45e2e190689b9f2e03e6b491ad8ab45

replay commented:
FYI force-pushing to rebase onto main because https://github.com/grafana/mimir/pull/2295#discussion_r930086446

replay commented on pkg/distributor/distributor.go:
```diff
@@ -908,18 +893,41 @@ func (d *Distributor) PushWithCleanup(ctx context.Context, req *mimirpb.WriteReq
 	return &mimirpb.WriteResponse{}, firstPartialErr
 }
 
+// forwardingReq returns a forwarding request if one is necessary, given the user ID.
```
Thx! 
Fixed in 1f86587e23e0b3fea2bc0137c3514dc7d42021e9

replay commented:
> I see that forwardSamples is called only from Push, but the method that API uses is PushWithCleanup.
>
> So this basically does not forward samples ingested through the API (Push is only used by the ruler as far as I can see).
>
> Is that a bug?

You're right 🤦‍♂️! Looks like this is lacking an integration test.

replay commented on pkg/distributor/distributor.go:
```diff
@@ -908,18 +893,41 @@ func (d *Distributor) PushWithCleanup(ctx context.Context, req *mimirpb.WriteReq
+// forwardingReq returns a forwarding request if one is necessary, given the user ID.
+// if no forwarding request is necessary it returns nil.
+func (d *Distributor) forwardSamples(ctx context.Context, req *mimirpb.WriteRequest, userID string, cleanup func()) (*mimirpb.WriteResponse, error) {
+	forwardingRules := d.limits.ForwardingRules(userID)
```
yeah, there's currently a bug because I only hooked it up with the `Push` method but actually the API uses `PushWithCleanup` ([Oleg's comment](https://github.com/grafana/mimir/pull/2295/#issuecomment-1195556150)). I'll fix that.

replay replied:
I changed this part completely.

I copied a lot of the ideas from Dimitar's [PR](https://github.com/grafana/mimir/pull/2419), basically the `PushWithCleanup` method now gets wrapped with an optional wrapper in the API, this should also lead to a more consistent solution once this PR and Dimitar's PR are merged: 

https://github.com/grafana/mimir/blob/fdfab1479188d4843299e3ff48c910906029c91c/pkg/api/api.go#L239-L245

replay commented on pkg/distributor/distributor.go:
```diff
@@ -436,6 +441,10 @@ func (d *Distributor) starting(ctx context.Context) error {
 }
 
 func (d *Distributor) running(ctx context.Context) error {
+	if d.cfg.Forwarding.Enabled && d.forwarder != nil {
```
great idea, thanks!

replay replied:
done, thanks again for the suggestion, this is definitely the better way to do it: https://github.com/grafana/mimir/blob/1373bff99be70bbb9433ce26b9b53aac6eb28b21/pkg/distributor/distributor.go#L382-L388

replay commented on pkg/distributor/forwarding/forwarding.go:
```diff
@@ -96,185 +74,260 @@ func NewForwarder(reg prometheus.Registerer, cfg Config) Forwarder {
+
+	tsByTargets := f.pools.getTsByTargets()
+	for tsSliceReadIdx, ts := range tsSlice {
+		ingest := false
```
good idea, done here: https://github.com/grafana/mimir/commit/05c7e94e808220b52a0fb58db8c61ec5020db1e0

replay commented:
I think I addressed all the comments, @colega and @56quarters if you get a chance it would be great if you could take another 👀 🙏 .

The messages about "Helm <> Jsonnet Diff" seem to be because whatever job is posting them doesn't seem to support generics 🤦‍♂️

colega commented on pkg/distributor/forwarding/forwarding.go:
```diff
@@ -96,185 +76,286 @@ func NewForwarder(reg prometheus.Registerer, cfg Config) Forwarder {
+func (r *request) do() {
+	defer r.cleanup()
+
+	protoBufBytes := (*r.pools.getProtobuf())[:0]
+	defer r.pools.putProtobuf(&protoBufBytes)
```
Non-blocking: I wonder if by saving the original reference we can avoid creating a new one.

Right now `r.pools.getProtobuf()` gives you a location in memory from where you're copying the slice header into `protoBufBytes` and throwing away the reference. However `protoBufBytes` also escapes to heap since you're returning a reference in the defer.

What if we do something like:
```suggestion
	protoBufBytesRef := r.pools.getProtobuf()
	protoBufBytes := (*protoBufBytesRef)[:0]
	defer func() {
		*protoBufBytesRef = protoBufBytes // just in case we increased its capacity
		r.pools.putProtobuf(protoBufBytesRef)
	}()
```

replay replied:
interesting idea!

I benchmarked it, the results were a bit inconclusive because I don't understand why only one of the two benchmark cases that are performing forwarding requests see an improvement, but if it "might" improve the allocations then I think we can apply this suggestion because the down-side that comes from the additional complexity is minimal (I don't think it's very complicated).

```
replay@mauro-workstation:~/go/src/github.com/grafana/mimir$ benchstat -delta-test none ./before ./after
name                                                      old time/op    new time/op    delta
RemoteWriteForwarding/forwarding_and_ingesting-4            1.66ms ± 0%    1.70ms ± 0%   +2.63%
RemoteWriteForwarding/forwarding_and_no_ingesting-4         1.76ms ± 0%    1.90ms ± 0%   +7.51%
RemoteWriteForwarding/no_forwarding_and_only_ingesting-4    1.44ms ± 0%    1.47ms ± 0%   +2.39%

name                                                      old alloc/op   new alloc/op   delta
RemoteWriteForwarding/forwarding_and_ingesting-4            2.43kB ± 0%    2.49kB ± 0%   +2.43%
RemoteWriteForwarding/forwarding_and_no_ingesting-4         2.53kB ± 0%    2.79kB ± 0%  +10.35%
RemoteWriteForwarding/no_forwarding_and_only_ingesting-4      190B ± 0%      192B ± 0%   +1.05%

name                                                      old allocs/op  new allocs/op  delta
RemoteWriteForwarding/forwarding_and_ingesting-4              29.0 ± 0%      28.0 ± 0%   -3.45%
RemoteWriteForwarding/forwarding_and_no_ingesting-4           29.0 ± 0%      29.0 ± 0%    0.00%
RemoteWriteForwarding/no_forwarding_and_only_ingesting-4      5.00 ± 0%      5.00 ± 0%    0.00%
```

colega commented on pkg/distributor/distributor.go:
```diff
@@ -598,26 +609,69 @@ func (d *Distributor) validateSeries(nowt time.Time, ts mimirpb.PreallocTimeseri
+		}
+
+		var errCh <-chan error
+		req.Timeseries, errCh = d.forwardSamples(ctx, userID, req.Timeseries)
+		resp, nextErr := next(ctx, req, cleanup)
```
The docs on your `forwarder.Forward` method now say:

> The slice of time series which gets passed into this function must not be returned to the pool by the caller

However, this `next(..., cleanup())` call will most likely return `req.Timeseries` back to the pool, right?

colega replied:
It will do here:

https://github.com/grafana/mimir/blob/4cc0bb1670a7b69912d40c4f6185a3ecd415c1ab/pkg/util/push/push.go#L84-L87

colega replied:
Discussed offline: this works because we reuse the updated `req.Timestamp` here, not the one we've created.

replay replied:
marking as resolved

colega commented on pkg/api/api.go:
```diff
@@ -239,7 +239,7 @@ func (a *API) RegisterRuntimeConfig(runtimeConfigHandler http.HandlerFunc) {
 	distributorpb.RegisterDistributorServer(a.server.GRPC, d)
 
-	wrappedDistributor := a.cfg.wrapDistributorPush(d)
+	wrappedDistributor := d.PrePushForwardingMiddleware(a.cfg.wrapDistributorPush(d))
```
I think that we need to update the docs on the `cfg.DistributorPushWrapper` now stating that this wrapper doesn't get all the pushed series, but only the ones that are going to be ingested throught he ingesters.

(I also would re-think twice whether that's what we want :thinking:)

replay replied:
> I think that we need to update the docs on the cfg.DistributorPushWrapper now stating that this wrapper doesn't get all the pushed series, but only the ones that are going to be ingested throught he ingesters.

Thx, I'll update the docs

> (I also would re-think twice whether that's what we want 🤔)

You're right, afaict this is used by business intelligence, I'll reach out to them after merging this about how they want this to be handled. Keep in mind that the aggregation results will go through another distributor, at which point the aggregation results would then be accounting for by the business intelligence wrapper.

replay replied:
fixed https://github.com/grafana/mimir/pull/2295/commits/e551b42e82a069c96580c1993c867b05ed9601fb

56quarters commented on pkg/distributor/forwarding/forwarding.go:
```diff
@@ -96,185 +76,286 @@ func NewForwarder(reg prometheus.Registerer, cfg Config) Forwarder {
-	ctx    context.Context
-	client *http.Client
-	pools  *pools
+func (f *forwarder) loop(ctx context.Context) error {
```
nit: If the only thing the "loop" method does you can `services.NewIdleService(start, stop)`

replay replied:
thanks! fixed c4b39b0c7e978d6a50e75a79409d82104ff1510e

56quarters reviewed (approved):
LGTM modulo any feedback from @colega

replay commented:
I will have to fix the broken `Helm <> Jsonnet Diff` test in a later follow-up PR

replay commented:
I deployed this branch into our dev cell and I noticed a problem: It seems that some label values of forwarded time series get corrupted. I think this is due to our usage of "yoloStrings" in mimirpb, to fix this we'll have to copy strings by value :( working on it.

replay commented:
@oleg and @56quarters  I pushed another commit which fixes the issue I described in my last comment, PTAL: https://github.com/grafana/mimir/pull/2295/commits/6ed53a31d9325a7dbdc6c06b2747e8f1564922ac

I don't like how complicated it is 😞 , it is pooling the byte slices which are backing the yolo strings created in `mimirpb`. If you have an idea how we can simplify this I would be very happy to hear it. Obviously, one way to simplify it would be to just use "normal" strings in `mimirpb.DeepCopyTimeseries()`, however this would increase the number of allocated objects by roughly about `7500%` according to the included benchmark, the current solution only increases the allocations by less than `10%`:

```
$ benchstat -delta-test none ./before after4
name                                                      old time/op    new time/op    delta
RemoteWriteForwarding/forwarding_and_ingesting-4            1.65ms ± 0%    1.74ms ± 0%   +5.13%
RemoteWriteForwarding/forwarding_and_no_ingesting-4         1.70ms ± 0%    1.81ms ± 0%   +6.13%
RemoteWriteForwarding/no_forwarding_and_only_ingesting-4    1.45ms ± 0%    1.38ms ± 0%   -4.65%

name                                                      old alloc/op   new alloc/op   delta
RemoteWriteForwarding/forwarding_and_ingesting-4            2.19kB ± 0%    2.52kB ± 0%  +15.15%
RemoteWriteForwarding/forwarding_and_no_ingesting-4         2.43kB ± 0%    2.60kB ± 0%   +7.08%
RemoteWriteForwarding/no_forwarding_and_only_ingesting-4      195B ± 0%      195B ± 0%    0.00%

name                                                      old allocs/op  new allocs/op  delta
RemoteWriteForwarding/forwarding_and_ingesting-4              27.0 ± 0%      29.0 ± 0%   +7.41%
RemoteWriteForwarding/forwarding_and_no_ingesting-4           28.0 ± 0%      29.0 ± 0%   +3.57%
RemoteWriteForwarding/no_forwarding_and_only_ingesting-4      5.00 ± 0%      5.00 ± 0%    0.00%
```

replay commented:
@colega had a great idea how I can simplify the code by moving the yolo slice pool into `mimirpb`. This makes the code significantly easier to understand 🎉. I implemented it in the commit: https://github.com/grafana/mimir/pull/2295/commits/f10adf0290deebdde04664fe56f07b425aca81e6

The benchmarks are also looking good, comparing `9f74614` (when I noticed the yolo string issue) against `f10adf0`:

```
$ benchstat -delta-test none ./9f74614 ./f10adf0
name                                                      old time/op    new time/op    delta
RemoteWriteForwarding/forwarding_and_ingesting-4            1.76ms ± 0%    2.00ms ± 0%  +13.60%
RemoteWriteForwarding/forwarding_and_no_ingesting-4         1.87ms ± 0%    2.01ms ± 0%   +7.73%
RemoteWriteForwarding/no_forwarding_and_only_ingesting-4    1.48ms ± 0%    1.46ms ± 0%   -1.52%

name                                                      old alloc/op   new alloc/op   delta
RemoteWriteForwarding/forwarding_and_ingesting-4            2.29kB ± 0%    2.78kB ± 0%  +21.40%
RemoteWriteForwarding/forwarding_and_no_ingesting-4         2.45kB ± 0%    2.89kB ± 0%  +17.99%
RemoteWriteForwarding/no_forwarding_and_only_ingesting-4      199B ± 0%      192B ± 0%   -3.52%

name                                                      old allocs/op  new allocs/op  delta
RemoteWriteForwarding/forwarding_and_ingesting-4              27.0 ± 0%      29.0 ± 0%   +7.41%
RemoteWriteForwarding/forwarding_and_no_ingesting-4           28.0 ± 0%      29.0 ± 0%   +3.57%
RemoteWriteForwarding/no_forwarding_and_only_ingesting-4      5.00 ± 0%      5.00 ± 0%    0.00%
```
